doc/tags_param.md

```diff
@@ -0,0 +1,248 @@
+
+See examples of `BaseTag` in [`SGObject.h`](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/base/SGObject.h).
+
+#### Any
+
+`Any` allows to store objects of arbitrary types in a type agnostic way. This makes it possible to store a variety of types in the `map`.
+
+```cpp
+int32_t integer = 5;                // integer
+GaussianKernel gkernel(5);          // gaussian kernel object
+auto any_integer = Any(integer);    // any object for integer
+auto any_kernel = Any(gkernel);     // any object of gaussian kernel
+```
+
+`Any` class has `BaseAnyPolicy` which is an interface for a policy to store a value. The value can be any data like primitive data-types, shogun objects, etc. and the policy defines how to handle this data. It works with a provided memory region and is able to set value, clear it and return the type-name as string. There are two derived classes of `BaseAnyPolicy` class:
+- `NonOwningValueAnyPolicy`: This uses external pointer in non-owning fashion (the pointer is never deleted by `Any` object) and new values are stored directly by the provided pointer.
+- `PointerValueAnyPolicy`: Unlike the above policy, this policy doesn't use external pointer to store values. But uses void pointers in owning fashion (the pointer is deleted by `Any` destructor).
+
+Now we look at an example using the two above discussed policies:
+```cpp
+// By default Any object uses PointerValueAnyPolicy
+int32_t value = 5;
+auto owning_any = Any(value);
+value = 6;
+auto owning_any_val = recall_type<int32_t>(owning_any);
+// owning_any_val=5 is not equal to value=6
+
+// Now if we use NonOwningValueAnyPolicy
+auto non_owning_any = Any::non_owning(&value);
+value = 7;
+auto non_owning_any_val = recall_type<int32_t>(non_owning_any);
+// non_owning_any_val=7 is equal to value=7
+```
+
+`Tag`, `BaseTag` and `Any` can be found in [`src/shogun/lib`](https://github.com/shogun-toolbox/shogun/tree/develop/src/shogun/lib).
+
+#### register_param() / register_member()
+
+While defining a new Shogun class, `register_param()` or `register_member()` should be used in the constructor to register parameters and member variables respectively. This would allow the parameters or class member variables' values to be queried or modified using `gets()` / `sets()`. Registering members and parameters are required to prevent Shogun users to modify / query only the registered variables and not any arbitrary variables. The new Shogun class should also inherit `CSGObject` as these two functions are `protected` in `CSGObject`. Let's look at an example to make this more concrete.
```

Comment:
I have explained this in `Any`'s `NonOwningValueAnyPolicy` section so didn't mention it again.